chainermn/communicators/_memory_utility.py

```diff
@@ -94,23 +94,32 @@ def extract_params_set_grad(model):
             if param.grad is not None]
 
 
-def pack_params(params, itemsize, attr_name, buffer, stream=None):
+def pack_params(params, itemsize, attr_name, buffer, stream=None, transfer_dtype=None):
     offset = 0
     for param in params:
         v = getattr(param, attr_name)
+        if None is not transfer_dtype:
+            v = v.astype(transfer_dtype)
```

Comment:
Could you only run it if the `dtype` of `v`  and `transfer_dtype` are different?
`pack_params()`  must be able to run asynchronously with cuda stream.